File: sdk/src/state.ts
import { PublicKey } from '@solana/web3.js';
import { deserialize, serialize } from 'borsh';
import { RobotData, CredentialData, CommandLogData, RobotStatus, PermissionLevel, CommandType } from './types';

class RobotAccount {
  isInitialized: boolean;
  owner: Uint8Array;
  authority: Uint8Array;
  status: number;
  robotId: Uint8Array;
  merkleRoot: Uint8Array;
  lastCommandTimestamp: bigint;
  totalCommandsExecuted: bigint;
  activeOperators: Uint8Array[];
  maxOperators: number;
  emergencyStop: boolean;
  metadataUri: string;

  constructor(fields: any) {
    this.isInitialized = fields.isInitialized;
    this.owner = fields.owner;
    this.authority = fields.authority;
    this.status = fields.status;
    this.robotId = fields.robotId;
    this.merkleRoot = fields.merkleRoot;
    this.lastCommandTimestamp = fields.lastCommandTimestamp;
    this.totalCommandsExecuted = fields.totalCommandsExecuted;
    this.activeOperators = fields.activeOperators;
    this.maxOperators = fields.maxOperators;
    this.emergencyStop = fields.emergencyStop;
    this.metadataUri = fields.metadataUri;
  }

  static schema = new Map([
    [
      RobotAccount,
      {
        kind: 'struct',
        fields: [
          ['isInitialized', 'u8'],
          ['owner', [32]],
          ['authority', [32]],
          ['status', 'u8'],
          ['robotId', [32]],
          ['merkleRoot', [32]],
          ['lastCommandTimestamp', 'i64'],
          ['totalCommandsExecuted', 'u64'],
          ['activeOperators', [[32]]],
          ['maxOperators', 'u8'],
          ['emergencyStop', 'u8'],
          ['metadataUri', 'string'],
        ],
      },
    ],
  ]);

  static decode(data: Buffer): RobotData {
    const decoded = deserialize(this.schema, RobotAccount, data) as RobotAccount;
    return {
      isInitialized: decoded.isInitialized !== 0,
      owner: new PublicKey(decoded.owner),
      authority: new PublicKey(decoded.authority),
      status: decoded.status as RobotStatus,
      robotId: Buffer.from(decoded.robotId),
      merkleRoot: Buffer.from(decoded.merkleRoot),
      lastCommandTimestamp: Number(decoded.lastCommandTimestamp),
      totalCommandsExecuted: Number(decoded.totalCommandsExecuted),
      activeOperators: decoded.activeOperators.map((op) => new PublicKey(op)),
      maxOperators: decoded.maxOperators,
      emergencyStop: decoded.emergencyStop !== 0,
      metadataUri: decoded.metadataUri,
    };
  }
}

class CredentialAccount {
  isInitialized: boolean;
  owner: Uint8Array;
  robot: Uint8Array;
  permissionLevel: number;
  validFrom: bigint;
  validUntil: bigint;
  revoked: boolean;
  credentialHash: Uint8Array;
  issuer: Uint8Array;

  constructor(fields: any) {
    this.isInitialized = fields.isInitialized;
    this.owner = fields.owner;
    this.robot = fields.robot;
    this.permissionLevel = fields.permissionLevel;
    this.validFrom = fields.validFrom;
    this.validUntil = fields.validUntil;
    this.revoked = fields.revoked;
    this.credentialHash = fields.credentialHash;
    this.issuer = fields.issuer;
  }

  static schema = new Map([
    [
      CredentialAccount,
      {
        kind: 'struct',
        fields: [
          ['isInitialized', 'u8'],
          ['owner', [32]],
          ['robot', [32]],
          ['permissionLevel', 'u8'],
          ['validFrom', 'i64'],
          ['validUntil', 'i64'],
          ['revoked', 'u8'],
          ['credentialHash', [32]],
          ['issuer', [32]],
        ],
      },
    ],
  ]);

  static decode(data: Buffer): CredentialData {
    const decoded = deserialize(this.schema, CredentialAccount, data) as CredentialAccount;
    return {
      isInitialized: decoded.isInitialized !== 0,
      owner: new PublicKey(decoded.owner),
      robot: new PublicKey(decoded.robot),
      permissionLevel: decoded.permissionLevel as PermissionLevel,
      validFrom: Number(decoded.validFrom),
      validUntil: Number(decoded.validUntil),
      revoked: decoded.revoked !== 0,
      credentialHash: Buffer.from(decoded.credentialHash),
      issuer: new PublicKey(decoded.issuer),
    };
  }
}

class CommandLogAccount {
  isInitialized: boolean;
  robot: Uint8Array;
  executor: Uint8Array;
  commandType: number;
  timestamp: bigint;
  parameters: Uint8Array;
  success: boolean;
  errorCode: number;

  constructor(fields: any) {
    this.isInitialized = fields.isInitialized;
    this.robot = fields.robot;
    this.executor = fields.executor;
    this.commandType = fields.commandType;
    this.timestamp = fields.timestamp;
    this.parameters = fields.parameters;
    this.success = fields.success;
    this.errorCode = fields.errorCode;
  }

  static schema = new Map([
    [
      CommandLogAccount,
      {
        kind: 'struct',
        fields: [
          ['isInitialized', 'u8'],
          ['robot', [32]],
          ['executor', [32]],
          ['commandType', 'u8'],
          ['timestamp', 'i64'],
          ['parameters', ['u8']],
          ['success', 'u8'],
          ['errorCode', 'u32'],
        ],
      },
    ],
  ]);

  static decode(data: Buffer): CommandLogData {
    const decoded = deserialize(this.schema, CommandLogAccount, data) as CommandLogAccount;
    return {
      isInitialized: decoded.isInitialized !== 0,
      robot: new PublicKey(decoded.robot),
      executor: new PublicKey(decoded.executor),
      commandType: decoded.commandType as CommandType,
      timestamp: Number(decoded.timestamp),
      parameters: Buffer.from(decoded.parameters),
      success: decoded.success !== 0,
      errorCode: decoded.errorCode,
    };
  }
}

export { RobotAccount, CredentialAccount, CommandLogAccount };

File: program/src/state.rs
use borsh::{BorshDeserialize, BorshSerialize};
use solana_program::{
    pubkey::Pubkey,
    program_pack::{IsInitialized, Sealed},
    clock::UnixTimestamp,
};

#[derive(BorshSerialize, BorshDeserialize, Debug, Clone, PartialEq)]
pub enum RobotStatus {
    Offline,
    Idle,
    Active,
    Executing,
    Error,
    Maintenance,
}

#[derive(BorshSerialize, BorshDeserialize, Debug, Clone, PartialEq)]
pub enum PermissionLevel {
    None,
    Observer,
    Operator,
    Administrator,
    Owner,
}

#[derive(BorshSerialize, BorshDeserialize, Debug, Clone)]
pub struct Robot {
    pub is_initialized: bool,
    pub owner: Pubkey,
    pub authority: Pubkey,
    pub status: RobotStatus,
    pub robot_id: [u8; 32],
    pub merkle_root: [u8; 32],
    pub last_command_timestamp: UnixTimestamp,
    pub total_commands_executed: u64,
    pub active_operators: Vec<Pubkey>,
    pub max_operators: u8,
    pub emergency_stop: bool,
    pub metadata_uri: String,
}

impl Robot {
    pub const LEN: usize = 1 + 32 + 32 + 1 + 32 + 32 + 8 + 8 + (32 * 10) + 1 + 1 + 256;
    
    pub fn new(
        owner: Pubkey,
        authority: Pubkey,
        robot_id: [u8; 32],
        merkle_root: [u8; 32],
        metadata_uri: String,
    ) -> Self {
        Self {
            is_initialized: true,
            owner,
            authority,
            status: RobotStatus::Idle,
            robot_id,
            merkle_root,
            last_command_timestamp: 0,
            total_commands_executed: 0,
            active_operators: Vec::new(),
            max_operators: 10,
            emergency_stop: false,
            metadata_uri,
        }
    }
}

impl Sealed for Robot {}

impl IsInitialized for Robot {
    fn is_initialized(&self) -> bool {
        self.is_initialized
    }
}

#[derive(BorshSerialize, BorshDeserialize, Debug, Clone)]
pub struct Credential {
    pub is_initialized: bool,
    pub owner: Pubkey,
    pub robot: Pubkey,
    pub permission_level: PermissionLevel,
    pub valid_from: UnixTimestamp,
    pub valid_until: UnixTimestamp,
    pub revoked: bool,
    pub credential_hash: [u8; 32],
    pub issuer: Pubkey,
}

impl Credential {
    pub const LEN: usize = 1 + 32 + 32 + 1 + 8 + 8 + 1 + 32 + 32;
    
    pub fn new(
        owner: Pubkey,
        robot: Pubkey,
        permission_level: PermissionLevel,
        valid_from: UnixTimestamp,
        valid_until: UnixTimestamp,
        credential_hash: [u8; 32],
        issuer: Pubkey,
    ) -> Self {
        Self {
            is_initialized: true,
            owner,
            robot,
            permission_level,
            valid_from,
            valid_until,
            revoked: false,
            credential_hash,
            issuer,
        }
    }
    
    pub fn is_valid(&self, current_timestamp: UnixTimestamp) -> bool {
        !self.revoked
            && current_timestamp >= self.valid_from
            && current_timestamp <= self.valid_until
    }
}

impl Sealed for Credential {}

impl IsInitialized for Credential {
    fn is_initialized(&self) -> bool {
        self.is_initialized
    }
}

#[derive(BorshSerialize, BorshDeserialize, Debug, Clone)]
pub struct CommandLog {
    pub is_initialized: bool,
    pub robot: Pubkey,
    pub executor: Pubkey,
    pub command_type: CommandType,
    pub timestamp: UnixTimestamp,
    pub parameters: Vec<u8>,
    pub success: bool,
    pub error_code: u32,
}

impl CommandLog {
    pub const LEN: usize = 1 + 32 + 32 + 1 + 8 + 256 + 1 + 4;
}

#[derive(BorshSerialize, BorshDeserialize, Debug, Clone, PartialEq)]
pub enum CommandType {
    Move,
    Rotate,
    Grab,
    Release,
    EmergencyStop,
    Reset,
    Calibrate,
    UpdateConfig,
    Custom,
}











